sql: sequences: respect MINVALUE and MAXVALUE options

vilterp opened the pull request:
Give an error if the sequence exceeds the min or max value, including handling integer overflows. Validate start w.r.t. bounds settings.

PG allows a `CYCLE` option which makes the value go to the `START` value when it hits the min or max, but leaving that out for now.

Commit:
sql: validate CREATE|ALTER SEQUENCE wrt bounds vs start

Commit:
sql: sequences: basic bound checks (without cycle or overflow)

Commit:
sql: fixup: fix mistaken doc comments

Commit:
sql: sequences: handle integer overflows

Commit:
sql: sequences: check sequence bounds in setval()

previously setval() allowed you to set a value outside of the sequence's
MINVALUE and MAXVALUE.

Release note: sql: check sequence MINVALUE/MAXVALUE in setval()

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20973)
<!-- Reviewable:end -->

justinj commented:
Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/alter_sequence.go, line 157 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21jqMLqXbACqaqWd7Y:-L21jqMLqXbACqaqWd7Z:b-g6jvt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/alter_sequence.go#L157)):*
> ```Go
> 
> 	// If start option not specified, set it to MinValue (for ascending sequences)
> 	// and Minvalue (for descending sequences).
> ```

`s/Minvalue/MaxValue/`

---

*[pkg/sql/sequence.go, line 138 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21ftE7IDO2TOO5xO2Z:-L21ftE7IDO2TOO5xO2_:bcskb9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/sequence.go#L138)):*
> ```Go
> 	}
> 	seqValueKey := keys.MakeSequenceKey(uint32(descriptor.ID))
> 	// TODO(vilterp): not supposed to use Put here
> ```

Can you explain why this is the case?

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 338 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21gWorO-SVvjXSKVxS:-L21gWorO-SVvjXSKVxT:b-2k06jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/logictest/testdata/logic_test/sequences#L338)):*
> ```
> 11
> 
> # Setval doesn't let you set values outside the bounds.
> ```

nit: other references to function names in this file seem to be lowercase

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 374 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21geAywzsz3tWgviD8:-L21geAywzsz3tWgviD9:b-55pkm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/logictest/testdata/logic_test/sequences#L374)):*
> ```
> 10
> 
> statement error pgcode 2200H pq: nextval\(\): reached maximum value of sequence "limit_test" \(10\)
> ```

maybe select `currval` here to see that it hasn't increased?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/alter_sequence.go, line 157 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21jqMLqXbACqaqWd7Y:-L2LxNsT34y5MH76e4c4:bdjdqqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/alter_sequence.go#L157)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`s/Minvalue/MaxValue/`
</blockquote></details>

Done. Also, moved this function from `alter_sequence.go` to `sequence.go`, since it's called from both `create_sequence.go` and `alter_sequence.go`.

---

*[pkg/sql/sequence.go, line 138 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21ftE7IDO2TOO5xO2Z:-L2LyF3IikKURS6FHxLM:b-46gpda) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/sequence.go#L138)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Can you explain why this is the case?
</blockquote></details>

Comments on the Increment operation say that you're not supposed to not mix using `increment` and `put` on a key https://github.com/cockroachdb/cockroach/blob/09013e0afadcc3f890dda8cb60ccb5ba5f3f533b/pkg/roachpb/api.proto#L183-L185

The comments don't say why this would be a problem (and it does work). You can use `Inc(desired_val - curr_val)` to achieve the same result, except this wouldn't work if the difference was really big — e.g. if someone was trying to reset their sequence from `math.MaxInt64` to `math.MinInt64`, `desired - curr` overflows to 1.

@petermattis, how big of a deal is it to use `Put` here, when we usually use `Inc` on this key?

(Note that this could be fixed in a separate PR; just added todo here when I saw it)

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 338 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21gWorO-SVvjXSKVxS:-L2Memt0qWK7J5TS1Lg4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/logictest/testdata/logic_test/sequences#L338)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: other references to function names in this file seem to be lowercase
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/sequences, line 374 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21geAywzsz3tWgviD8:-L2MfjP6656IldLoc1OW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/logictest/testdata/logic_test/sequences#L374)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

maybe select `currval` here to see that it hasn't increased?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973)*
<!-- Sent from Reviewable.io -->

justinj commented:
LGTM modulo someone addressing the Inc/Put thing

---

Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sequence.go, line 138 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-L21ftE7IDO2TOO5xO2Z:-L2QOJzn2WTx9ZPpM_JV:bmjk3b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef00439b93d69da23d52f9cdc4060b07a31f2a8/pkg/sql/sequence.go#L138)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Comments on the Increment operation say that you're not supposed to not mix using `increment` and `put` on a key https://github.com/cockroachdb/cockroach/blob/09013e0afadcc3f890dda8cb60ccb5ba5f3f533b/pkg/roachpb/api.proto#L183-L185

The comments don't say why this would be a problem (and it does work). You can use `Inc(desired_val - curr_val)` to achieve the same result, except this wouldn't work if the difference was really big — e.g. if someone was trying to reset their sequence from `math.MaxInt64` to `math.MinInt64`, `desired - curr` overflows to 1.

@petermattis, how big of a deal is it to use `Put` here, when we usually use `Inc` on this key?

(Note that this could be fixed in a separate PR; just added todo here when I saw it)
</blockquote></details>

I see, that seems like a pretty big issue but I'll defer to someone with more understanding of that area. If you decide to push that to another PR it would probably be good to include some of what you wrote here in the TODO.

Re: your overflow thing, is that not okay? If `desired - curr` is 1 then that would overflow `math.MaxInt64` to `math.MinInt64`, wouldn't it? Or do these keys not overflow in that way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20973#-:-L2QPVfZ7J9O4YkXyRNF:b-f9bd50)*
<!-- Sent from Reviewable.io -->
